Mathlib/RingTheory/Kaehler/JacobiZariski.lean

```diff
@@ -0,0 +1,559 @@
+    have : mk ⟨x, hx'⟩ = mk z := by
+      ext; simpa only [comp_vars, val_mk, Ideal.toCotangent_eq, sub_sub_cancel, pow_two]
+    rw [this, ← Submodule.restrictScalars_mem (Q.comp P).Ring, ← Submodule.mem_comap,
+      ← Submodule.span_singleton_le_iff_mem, ← Submodule.map_le_map_iff_of_injective
+      (f := Submodule.subtype _) Subtype.val_injective, Submodule.map_subtype_span_singleton,
+      Submodule.span_singleton_le_iff_mem]
+    refine (show Ideal.map (Q.toComp P).toAlgHom.toRingHom P.ker ≤ _ from ?_) hz
+    rw [Ideal.map_le_iff_le_comap]
+    rintro w hw
+    simp only [AlgHom.toRingHom_eq_coe, Ideal.mem_comap, RingHom.coe_coe,
+      Submodule.mem_map, Submodule.mem_comap, Submodule.restrictScalars_mem, Submodule.coe_subtype,
+      Subtype.exists, exists_and_right, exists_eq_right]
+    refine ⟨?_, Submodule.subset_span ⟨.mk ⟨w, hw⟩, ?_⟩⟩
+    · simp only [ker, RingHom.mem_ker, algebraMap_apply, Hom.algebraMap_toAlgHom]
+      rw [← algebraMap_apply, hw, map_zero]
+    · rw [map_mk]
+
+/-- Given `R[X] → S` and `S[Y] → T`, the cotangent space of `R[X][Y] → T` is isomorphic
+to the direct product of the cotangent space of `S[Y] → T` and `R[X] → S` (base changed to `T`). -/
+noncomputable
+def CotangentSpace.compCotangentSpace (Q : Generators.{w} S T) (P : Generators.{w'} R S) :
+    (Q.comp P).CotangentSpace ≃ₗ[T] Q.CotangentSpace × (T ⊗[S] P.CotangentSpace) :=
+  (Q.comp P).cotangentSpaceBasis.repr.trans
+    (Q.cotangentSpaceBasis.prod (P.cotangentSpaceBasis.baseChange T)).repr.symm
+
+lemma CotangentSpace.compCotangentSpace_symm_inr
+    (Q : Generators.{w} S T) (P : Generators.{w'} R S) :
+    (compCotangentSpace Q P).symm.toLinearMap ∘ₗ
+      LinearMap.inr T Q.CotangentSpace (T ⊗[S] P.CotangentSpace) =
+        (CotangentSpace.map (Q.toComp P)).liftBaseChange T := by
+  classical
+  apply (P.cotangentSpaceBasis.baseChange T).ext
+  intro i
+  apply (Q.comp P).cotangentSpaceBasis.repr.injective
+  ext j
+  simp only [compCotangentSpace, LinearEquiv.trans_symm, LinearEquiv.symm_symm,
+    Basis.baseChange_apply, LinearMap.coe_comp, LinearEquiv.coe_coe, LinearMap.coe_inr,
+    Function.comp_apply, LinearEquiv.trans_apply, Basis.repr_symm_apply, pderiv_X,
+    Basis.repr_linearCombination, LinearMap.liftBaseChange_tmul, one_smul, repr_map, toComp_val]
+  obtain (j | j) := j <;>
+    simp only [comp_vars, Basis.prod_repr_inr, Basis.baseChange_repr_tmul,
+      Basis.repr_self, Basis.prod_repr_inl, map_zero, Finsupp.coe_zero,
+      Pi.zero_apply, ne_eq, not_false_eq_true, Pi.single_eq_of_ne, Pi.single_apply,
+      Finsupp.single_apply, ite_smul, one_smul, zero_smul, Sum.inr.injEq,
+        RingHom.map_ite_one_zero, reduceCtorEq, ↓reduceIte]
+
+lemma CotangentSpace.compCotangentSpace_symm_zero
+    (Q : Generators.{w} S T) (P : Generators.{w'} R S) (x) :
+    (compCotangentSpace Q P).symm (0, x) =
+        (CotangentSpace.map (Q.toComp P)).liftBaseChange T x :=
+  DFunLike.congr_fun (compCotangentSpace_symm_inr Q P) x
```

Comment:
It "Used 42728 heartbeats, which is less than the current maximum of 200000." on my local machine though.

Reply:
I think this is not the `synthInstance` heartbeat count. What I did not realize though is that this just fails because `SMulCommClass.of_commMonoid` is not an instance. The lemmas that time out b/c of instance search come later.

Reply:
What do you think of moving
```
attribute [local instance 999999] Zero.toOfNat0 SemilinearMapClass.distribMulActionSemiHomClass
  SemilinearEquivClass.instSemilinearMapClass TensorProduct.addZeroClass AddZeroClass.toZero
```
to the lemmas that need this manual tweaking? We should fix this at some point, so then it is quicker to find where it is going wrong. My suggestion would be to move this into a section starting at L335 and ending at L389. And then add it to `H1Cotangent.δ_map` with `attribute [local instance] .... in`.
Maybe you can also add a comment saying "Note: these instances are needed to prevent instance search timeouts."